File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.Comparator;
import java.util.StringTokenizer;

public class Main {
	
    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        
        String s = br.readLine();
        ArrayDeque<Character> stack = new ArrayDeque<>();
        while (!s.equals(".")) {
        	stack.clear();
        	boolean wrong = false;
        	char[] chars = s.toCharArray();
        	for (char c : chars) {
        		if (c == '[' || c == '(') {
        			stack.offerLast(c);
        		} else if (c == ']') {
        			
        			if (stack.isEmpty() || stack.peekLast() != '[') {
        				wrong = true;
        				break;
        			} else {
        				stack.pollLast();
        			}
        		} else if (c == ')') {
        			if (stack.isEmpty() || stack.peekLast() != '(') {
        				wrong = true;
        				break;
        			} else {
        				stack.pollLast();
        			}
        		}
        	}
        	if (!stack.isEmpty()) wrong = true;
        	if (wrong) {
        		sb.append("no");
        	} else {
        		sb.append("yes");
        	}
        	sb.append("\n");
        	s = br.readLine();
        }
        System.out.println(sb.toString());
    }
}

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.py
import sys

s = sys.stdin.readline().rstrip()

while s != "." :
    stack = []
    sol = True
    for c in s :
        if c == '[' or c == '(' :
            stack.append(c)
        elif c == ')' :
            if len(stack) != 0  and stack[-1] == '(' :
                stack.pop()
            else :
                sol = False
                break
        elif c == ']':
            if len(stack) != 0 and stack[-1] == '[':
                stack.pop()
            else:
                sol = False
                break
    if len(stack) != 0 :
        sol = False
    if sol :
        print("yes")
    else :
        print("no")
    s = sys.stdin.readline().rstrip()